src/number.rs

```diff
@@ -1,47 +1,327 @@
+
+    fn mul(self, rhs: AbstractNumberType) -> Self::Output {
+        AbstractNumberType {
+            value: self.value * rhs.value,
+        }
+    }
+}
+
+// Div
+
+impl std::ops::Div for AbstractNumberType {
+    type Output = AbstractNumberType;
+
+    fn div(self, rhs: Self) -> Self::Output {
+        AbstractNumberType {
+            value: self.value / rhs.value,
+        }
     }
 }
 
-pub fn is_zero(x: &AbstractNumberType) -> bool {
-    x.sign() == Sign::NoSign
+impl<'a> std::ops::Div<&'a AbstractNumberType> for AbstractNumberType {
+    type Output = AbstractNumberType;
+
+    fn div(self, rhs: &'a AbstractNumberType) -> Self::Output {
+        AbstractNumberType {
+            value: self.value / rhs.value,
+        }
+    }
 }
 
-lazy_static! {
-    // default field modulus is goldilocks
-    static ref GOLDILOCKS_MOD : BigInt = BigInt::parse_bytes(b"FFFFFFFF00000001", 16).unwrap();
-    static ref FIELD_MOD : Mutex<BigInt> = Mutex::new(BigInt::parse_bytes(b"FFFFFFFF00000001", 16).unwrap());
+impl<'a> std::ops::Div for &'a AbstractNumberType {
+    type Output = AbstractNumberType;
+
+    fn div(self, rhs: Self) -> Self::Output {
+        AbstractNumberType {
+            value: self.value / rhs.value,
+        }
+    }
 }
 
-pub fn get_goldilocks_mod() -> BigInt {
-    GOLDILOCKS_MOD.clone()
+impl<'a> std::ops::Div<AbstractNumberType> for &'a AbstractNumberType {
+    type Output = AbstractNumberType;
+
+    fn div(self, rhs: AbstractNumberType) -> Self::Output {
+        AbstractNumberType {
+            value: self.value / rhs.value,
+        }
+    }
 }
 
-pub fn get_field_mod() -> BigInt {
-    FIELD_MOD.lock().unwrap().clone()
+// Rem
+
+impl std::ops::Rem for AbstractNumberType {
+    type Output = AbstractNumberType;
+
+    fn rem(self, rhs: Self) -> Self::Output {
+        (self.to_degree() % rhs.to_degree()).into()
+    }
 }
 
-pub fn set_field_mod(n: BigInt) {
-    *FIELD_MOD.lock().unwrap() = n;
+impl std::ops::BitAnd for AbstractNumberType {
+    type Output = AbstractNumberType;
+
+    fn bitand(self, rhs: Self) -> Self::Output {
+        (self.to_degree() & rhs.to_degree()).into()
+    }
+}
+
+impl std::ops::BitAndAssign for AbstractNumberType {
+    fn bitand_assign(&mut self, rhs: Self) {
+        use std::ops::BitAnd;
+        *self = self.bitand(rhs);
+    }
+}
+
+impl std::ops::BitOr for AbstractNumberType {
+    type Output = AbstractNumberType;
+
+    fn bitor(self, rhs: Self) -> Self::Output {
+        (self.to_degree() | rhs.to_degree()).into()
+    }
+}
+
+impl std::ops::BitOrAssign for AbstractNumberType {
+    fn bitor_assign(&mut self, rhs: Self) {
+        use std::ops::BitOr;
+        *self = self.bitor(rhs);
+    }
+}
+
+impl std::ops::BitXor for AbstractNumberType {
+    type Output = AbstractNumberType;
+
+    fn bitxor(self, rhs: Self) -> Self::Output {
+        (self.to_degree() ^ rhs.to_degree()).into()
+    }
+}
+
+impl std::ops::Shl<DegreeType> for AbstractNumberType {
```

Comment:
Do we really want to implement this for the finite field?

Reply:
I can inline it where it's used instead, sure